Core concept page (#1082)
novuhq/docs: content/docs/agents/get-started/core-concepts.mdx

A conversation can include multiple participants. The agent is one participant in the conversation, but it is not the conversation itself. This distinction matters when you inspect conversations in the dashboard or build agent behavior around conversation state. A Slack thread, email thread, or other provider-specific thread maps to a conversation in Novu automatically.

The conversations follow a simple lifecycle, a conversation starts as active when the first message arrives. Messages, actions, and reactions continue updating the conversation while it remains active. The conversation can be marked as resolved when the agent calls the resolve signal. When a conversation is resolved, the resolve handler fires and an optional summary is stored.

If a user sends a new message after a conversation has been resolved, the conversation automatically reopens so the interaction can continue.

### Participants and subscriber identity

Participants are the people, agents, or systems involved in a conversation. When a user messages your agent from a provider, Novu tries to resolve that platform user to a known subscriber. 

> If a match exists, the handler receives subscriber information such as the subscriber ID, name, or email. If no match exists, the user is tracked as a platform user and your handler should account for cases where subscriber data is unavailable. Once identity is resolved later, they are upgraded to a full subscriber and future messages include their subscriber data.

Reviewer: ```suggestion
- If a match exists, then the handler receives subscriber information such as the subscriber ID, name, or email.
- If no match exists, then the user is tracked as a platform user and your handler should account for cases where subscriber data is unavailable.

Once identity is resolved later, they are upgraded to a full subscriber and future messages include their subscriber data.
```

## The bridge surface

Novu exposes a consistent set of building blocks to your agent code. These are the same regardless of which provider the message came from or what brain you've plugged in.

### Event handlers

Event handlers are functions that respond to events in a conversation. Your agent can respond to four event types:
 
| Handler | When it runs | Common use case |
|---------|-------------|-----------------|
| `onMessage` | A user sends a message in the conversation | Process the message and reply |
| `onAction` | A user clicks a button or selects a value in an interactive card | Handle form submissions, button clicks, or dropdown selections |
| `onReaction` | A user adds or removes a reaction | Capture feedback or trigger a follow-up |
| `onResolve` | The conversation is marked as resolved | Clean up state, log analytics, or send a summary |

Handlers are where the communication layer connects to your application logic. For example, an `onMessage` handler receives the user's message, passes the conversation context to an LLM or custom function, and sends the response back through Novu.

> 

Reviewer: ```suggestion
```

### Context object

Each event handler receives a context object. The context object gives your handler the information it needs to understand the current event and respond to it. Depending on the event type, it can include:
* The incoming message.
* The current conversation state and metadata.
* The resolved subscriber (when available).
* Recent conversation history.
* Provider information.
* Platform-specific details, such as thread or channel identifiers.
* Methods for replying, updating metadata, triggering workflows, or resolving the conversation.
 
The context object is the single interface between Novu's communication layer and your agent logic. Your code doesn't need to talk directly to Slack, Microsoft Teams, WhatsApp, or email, everything flows through the context.

> 

Reviewer: ```suggestion
```

 
| Signal | What it does |
|--------|-------------|
| Metadata | Stores key-value data on the conversation that persists across messages |
| Trigger | Starts a Novu workflow from within the conversation |
| Resolve | Marks the conversation as resolved with an optional summary |
 
A useful way to think about the difference: replies communicate with the user, signals update the system around the conversation. A single handler invocation can do both, reply to the user, store a sentiment score in metadata, trigger an escalation workflow, and resolve the conversation.

Signals aren't sent immediately when you call them. They're queued in memory and batched together with your next reply call in a single request. If your handler finishes without calling reply handler, any pending signals are still sent automatically.

## Conversations and workflows

ACI's conversation model and Novu's workflows are designed to work together. A conversation can trigger a workflow.

> For example, a user chatting with an agent in Slack asks it to prepare a report and send it via email. The agent calls a trigger functtion to fire a Novu workflow that handles the email delivery using the same workflow infrastructure your team may already have in place.

Reviewer: ```suggestion
For example, a user chatting with an agent in Slack asks it to prepare a report and send it via email. The agent calls a trigger function to fire a Novu workflow that handles the email delivery using the same workflow infrastructure your team might already have in place.
```

A workflow can convert into a conversation. For example, you send a weekly digest notification to a user's email. The user replies to that email with a follow-up question, and that reply enters ACI as a new conversation with the agent. What started as a one-directional notification becomes a back-and-forth exchange.

This connection means ACI is not a replacement for Novu's workflow system — it's an extension of it. The structured, predictable nature of transactional workflows and the open-ended, freeform nature of agent conversations are complementary. The value is in the combination of the two within the same platform.

> 

Reviewer: ```suggestion
```
